Pal/src/host/Linux/db_object.c

```diff
@@ -185,10 +205,24 @@ int _DkObjectsWaitAny (int count, PAL_HANDLE * handleArray, uint64_t timeout,
                 !(HANDLE_HDR(hdl)->flags & ERROR(j)))
                 events |= POLLIN;
 
-            if ((HANDLE_HDR(hdl)->flags & WFD(j)) &&
-                !(HANDLE_HDR(hdl)->flags & WRITEABLE(j)) &&
-                !(HANDLE_HDR(hdl)->flags & ERROR(j)))
-                events |= POLLOUT;
+            /* DEP 4/2/18: Go ahead and check for POLLOUT even if 
+             * we have already cached the WRITEABLE property.
+             * It should go quickly.  Or, we could quit early.
+             */
+            if ((HANDLE_HDR(hdl)->flags & WFD(j))) {
+
+                if (!(HANDLE_HDR(hdl)->flags & WRITEABLE(j)) &&
+                    !(HANDLE_HDR(hdl)->flags & ERROR(j))) 
+                    events |= POLLOUT; 
+                else if ((!(HANDLE_HDR(hdl)->flags & ERROR(j)))
+                         && events && hdl->generic.fds[j] != PAL_IDX_POISON) {
+                    // We should be able to at least return that this handle
+                    // is writeable, if anyone cares.  We only need to return
+                    // one, so it is ok to set the last one.
+                    timeout = 0;
```

Comment:
The following is a simple test case failing on this branch's pal-Linux but succeeding natively. poll() is supposed to wait 5 seconds for the DNS reply, but Graphene's internal logic swallows that timeout.

I am actually wondering if the PAL API should be changed. Within DkObjectsWaitAny the application's intent of whether it is polling on a read or write is not apparent. DkObjectsWaitAny decides on whether to poll on read or write solely based on the state of the pal handle (unless I am missing something). Maybe that design should be revisited.
 
```
#include <assert.h>
#include <stdlib.h>
#include <stdio.h>
#include <strings.h>
#include <netdb.h>
#include <sys/epoll.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <arpa/inet.h>

int main(int argc, char* argv[]) {
    struct addrinfo *res;
    int ret = getaddrinfo("proxy.jf.intel.com", NULL, NULL, &res);
    assert(0 == ret);
    return 0;
}

```

Reply:
Does #249 address this issue?  I think there are other changes in the works to address this (e.g., #229).

Reply:
So I think this is fixed on master, sort of.  When I try this test locally, it fails immediatley with:
EAI_NONAME	  -2	/* NAME or SERVICE is unknown.  */

If I do it in Graphene master, with a restrictive manifest, I can manufacture this timeout after 5 seconds (errno 110 - ETIMEDOUT).  I'm guessing this particular url behaves differently outside of intel?

I think the right thing to do here is make a unit test out of the code you propose, and make sure this still passes once we rebase the current PR to master.  But I'd like the unit test to work the same on Linux and in Graphene.  Can you propose a url or variant that always times out?

In the interest of preserving this behavior, and since this looks like the last remaining issue on this PR, can I go ahead and squash/rebase/force push this?

Reply:
Just tried it on master, but still fails for the same reason. Have not tried with a machine outside of the Intel network, but changing the domain to "intel.com" shows the same behavior.

This is the code I used (stored in LibOS/shim/test/native/getaddrinfo.c)
```
#include <assert.h>
#include <netdb.h>
#include <stdlib.h>                                                                                                                                                                                                                    
#include <sys/types.h>
#include <sys/socket.h>

int main(int argc, char* argv[]) {
    struct addrinfo *res;
    int ret = getaddrinfo("intel.com", NULL, NULL, &res);
    assert(0 == ret);
    return 0;
}
```

LibOS/shim/test/native/getaddrinfo.manifest.template
```
loader.preload = file:$(SHIMPATH)
loader.env.LD_LIBRARY_PATH = /lib
loader.debug_type = inline
loader.syscall_symbol = syscalldb

fs.mount.lib.type = chroot
fs.mount.lib.path = /lib
fs.mount.lib.uri = file:$(LIBCDIR)

fs.mount.bin.type = chroot
fs.mount.bin.path = /bin
fs.mount.bin.uri = file:/bin

sys.brk.size = 32M
sys.stack.size = 4M

# allow to bind on port 8000
net.allow_bind.1 = 127.0.0.1:8000
# allow to connect to port 8000
net.allow_peer.1 = 127.0.0.1:8000

# sgx-related
sgx.trusted_files.ld = file:$(LIBCDIR)/ld-linux-x86-64.so.2
sgx.trusted_files.libc = file:$(LIBCDIR)/libc.so.6
sgx.trusted_files.libdl = file:$(LIBCDIR)/libdl.so.2
sgx.trusted_files.libnss1 = file:/lib/x86_64-linux-gnu/libnss_compat.so.2
sgx.trusted_files.libnss2 = file:/lib/x86_64-linux-gnu/libnss_files.so.2
sgx.trusted_files.libnss3 = file:/lib/x86_64-linux-gnu/libnss_nis.so.2
sgx.trusted_files.libnss4 = file:/lib/x86_64-linux-gnu/libnss_myhostname.so.2
sgx.trusted_files.libnsl = file:/lib/x86_64-linux-gnu/libnsl.so.1
sgx.trusted_files.libnssdns = file:$(LIBCDIR)/libnss_dns.so.2                                                                                                                                                                          
sgx.trusted_files.libresolv = file:$(LIBCDIR)/libresolv.so.2
```

I built this by doing in LibOS/shim/test/native

```
rm -f getaddrinfo.manifest{,.sgx,token,sig} ; make -k SGX=1 DEBUG=1 ; make -k SGX_RUN=1 DEBUG=1
```

I actually have a fix locally by extending #229. I should push this.

Reply:
Would it make sense to let this PR go forward and just do a separate PR that adds this fix (and the unit tests)?  I'm stoked that you have already fixed this.

Reply:
@thomasknauth Bump

Reply:
Yes, please go ahead an close this PR. I am getting my act together and finally submit my extension of #229 shortly.

Reply:
Thanks!